pallets/stellar-relay/src/lib.rs

```diff
@@ -565,27 +584,60 @@ pub mod pallet {
 
 				ensure!(node_id_found, Error::<T>::EnvelopeSignedByUnknownValidator);
 
+				// Check if all envelopes are using the same slot index
+				ensure!(
+					slot_index == envelope.statement.slot_index,
+					Error::<T>::EnvelopeSlotIndexMismatch
+				);
+
 				let signature_valid = verify_signature(envelope, &node_id, network);
 				ensure!(signature_valid, Error::<T>::InvalidEnvelopeSignature);
 			}
 
 			// Check if transaction set matches tx_set_hash included in the ScpEnvelopes
 			let expected_tx_set_hash = compute_non_generic_tx_set_content_hash(transaction_set)
-				.ok_or(Error::<T>::FailedToComputenonGenericTxSetContentHash)?;
+				.ok_or(Error::<T>::FailedToComputeNonGenericTxSetContentHash)?;
+
+			// We store the externalized value in a variable so that we can check if it's the same
+			// for all envelopes. We don't distinguish between externalized and confirmed values as
+			// it should be the same value regardless.
+			let mut externalized_value: Option<Value> = None;
+			let mut externalized_n_h: Option<u32> = None;
 
 			for envelope in envelopes.get_vec() {
-				match envelope.clone().statement.pledges {
-					ScpStatementPledges::ScpStExternalize(externalized_statement) => {
-						let tx_set_hash = Self::get_tx_set_hash(&externalized_statement)?;
-						ensure!(
-							tx_set_hash == expected_tx_set_hash,
-							Error::<T>::TransactionSetHashMismatch
-						);
-					},
+				let (value, n_h) = match envelope.clone().statement.pledges {
+					ScpStatementPledges::ScpStExternalize(externalized_statement) =>
+						(externalized_statement.commit.value, externalized_statement.n_h),
+					ScpStatementPledges::ScpStConfirm(confirmed_statement) =>
+						(confirmed_statement.ballot.value, confirmed_statement.n_h),
 					_ => return Err(Error::<T>::InvalidScpPledge),
+				};
+
+				// Check if the tx_set_hash matches the one included in the envelope
+				let tx_set_hash = Self::get_tx_set_hash(&value)?;
+				ensure!(
+					tx_set_hash == expected_tx_set_hash,
+					Error::<T>::TransactionSetHashMismatch
+				);
+
+				// Check if the externalized value is the same for all envelopes
+				if let Some(externalized_value) = &externalized_value {
+					ensure!(externalized_value == &value, Error::<T>::ExternalizedValueMismatch);
+				} else {
+					externalized_value = Some(value);
+				}
+
+				// Check if the externalized nh is the same for all envelopes
+				if let Some(externalized_nh) = &externalized_n_h {
+					ensure!(externalized_nh == &n_h, Error::<T>::ExternalizedNHMismatch);
+				} else {
+					externalized_n_h = Some(n_h);
 				}
```

Comment:
@TorstenStueber I added a check for this here, or are we talking about different things?

Reply:
That's correct.